ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PluginPrivacyController.java

```diff
@@ -119,9 +124,16 @@ public Transaction createPrivateMarkerTransaction(
     final BlockHeader blockHeader =
         blockchain.getBlockHeader(transactionLocation.getBlockHash()).orElseThrow();
 
-    final BytesValueRLPInput input = new BytesValueRLPInput(transaction.get().getPayload(), false);
+    final Optional<org.hyperledger.besu.plugin.data.PrivateTransaction> pluginPrivateTransaction =
+        privacyPluginService
+            .getPayloadProvider()
+            .getPrivateTransactionFromPayload(blockHeader.getNumber(), transaction.get());
```

Comment:
We definitely need blockNumber in the plugin for ~decrypting~ `getPrivateTransactionFromPayload`. You're right when we ~encrypt~ `generateMarkerPayload` the plugin can not know for certain. 

Let's say we have a simple scheme to use a different key every 1000 blocks. When you call `generateMarkerPayload` the plugin could ask the node for the block height (this would, of course, be increasing on a different thread so you can not use it as a guarantee) if the height is after 1000 use key A and if it's after 2000 use key B. It could also reject the transaction if the block height is too close to the crossover. 

When you come to ~decrypt~ `getPrivateTransactionFromPayload` you need the block number for the transaction. It's not in `Transaction` you could get it using `eth_getTransactionByHash`. 

I think it's useful, but happy to remove it as there are ways to get the info you need.

Reply:
I think we should remove it. In the generateMarkerPayload(privateTransaction, privacyUserId) you are not passing it in and in general I do not like it if something is only "about" right. I guess if you do want to rotate keys you could always do that based on time, or what sometimes is done, based on the amount of data that has been encrypted with a key. 
In case you rotate keys you would generally provide some kind of key id in the payload to enable the receiving side to find the right key.